[FLINK-15670][API] Provide a Kafka Source/Sink pair as KafkaShuffle

curcur opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request is to provide a Kafka Source/Sink pair to simulate a KafkaShuffle that can:
1. Read topics that are already partitioned by key and process them without partitioning them again (avoid shuffles); and
2. Use this to decompose the job into smaller jobs and independent pipelined regions that failover independently.

## Brief Change Log
  - Extend Datastream API to allow user-defined SinkFunction to manipulate watermark
  - Kafka Shuffle Producer change
    - Write timestamp and watermark information together with a record in Kafka
    - Use `KeyGroupRangeAssignment` to assign records to different Kafka partitions
  - Kafka Shuffle Consumer change, Kafka Fechter Change
    - Each consumer read partitions equal to the key group indices that it got assigned. For now, the number of partitions is enforced to equal to the number of consumer parallelism.
    - KafkaShuffleFechter encapsulates the logic of record emitting and watermark emitting. 

## Verifying this change
This change added tests:
1. End-2-end tests
2. partition assignment tests

Some other tests are missing:
1. Specific watermark tests
2. failover tests

## Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (don't know)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

curcur commented:
add Java Docs

curcur commented:
Hey @AHeise , Thank you so much for taking the time to review the code and providing feedback!

I think you are totally right: I have the same confusion when I first read the JIRA. That's why we re-define the scope of the issue.  Do you think it would be helpful (we two first) to chat offline syncing up a bit?

Then if we can involve Stephan after reaching some level of agreement? I will ping you offline.

curcur commented:
List some of the points here, just in case I forgot :-). We can chat details offline.

1. If I am understanding correctly (please ignore if I am not), the original idea is to make use of the partitioned data so if users consume data they do not do an extra "reinterpret". The problem is KeyedStream needs a specific way to decide key -> PartitionID (KeyGroupRangeAssignment.assignKeyToParallelOperator). If the Keyed data in Kafka does not follow this hash function, it has to do a repartition (keyBy). 

In other words, if the data is ingested from other systems, there is no easy way to guarantee the way an external system partitions data the same way Flink needs. I think that's why Stephan suggests providing a Kafka Writer as well, so the way how data is partitioned is controlled internally by us.

`reinterpretAsKeyedStream` does not need an extra shuffle (if that's the concern).

2. How people can reuse the data
We can provide a read API for people to read, that should not be difficult to do (without letting them worrying about watermark), they only need to provide a data schema.
The original persist API has a write and read separately. But I guess Stephan is more interested in the "Shuffle idea", so I wrapped them together to make the code a bit cleaner.

3. Why watermark is needed
If we treat this as a pure sink/source, well, you are right, the watermark is not needed. But it does not seem that useful. I think Stephan is more interested in taking this as a shuffle. Me as well, hmm. As a shuffle, watermark is not avoidable.

4. This is one way we design how to watermark is transitted. I agree that there are a lot of different ways (some of them I have actually tried, for example, timestamp). My guideline is to make the dependency on Kafka as small as possible. 

5. Some other concerns like why I have an extra sink function and operator class: to avoid the effects/changes of the current interface. SinkFunctions and Operators are broadly used, and I do not want to cause confusions to our users.

AHeise commented:
>     1. If I am understanding correctly (please ignore if I am not), the original idea is to make use of the data already partitioned. So that users do not do an extra "reinterpret".
> 
> 
> The problem is KeyedStream needs a specific way to decide key -> PartitionID `KeyGroupRangeAssignment.assignKeyToParallelOperator`. If the Keyed data in Kafka does not follow this hash function, it has to do a repartition (keyBy).
> 
> In other words, there is no easy way to guarantee the way an external system partitions data using the same way Flink uses. I think that's why Stephan suggests providing a Kafka Writer as well in the Jira. In this way how data is partitioned is controlled internally by us.
> 
> `reinterpretAsKeyedStream` itself does not need an extra shuffle (if that's the concern).

First off, I haven't fully considered of using Kafka as an external shuffle service. I know of plans to have persistent communication channels, which have a high conceptional overlap with `KafkaShuffle`. I fear that once these persistent channels come, it may confuse users when to use what.

Then, my second thought on `KafkaShuffle` is, why not directly provide it as an `ShuffleServiceFactory`? Do we want a more selective approach on the shuffle service or is the current interface not suitable (e.g., how to determine topic name)?

So while I'm still not convinced of the general direction (and I originally had something else in mind), I'm assuming for now that this `KafkaShuffle` already gives some value to users. Then the topic should be almost treated as an implementation detail.

>     1. How people can reuse the data
>        We can provide a read API for people to read, that should not be difficult to do (without letting them worrying about watermark), they only need to provide a data schema.

On the premise of having a pure shuffle service, I agree that this is an implementation detail. I was wondering if users of external systems may also use the data as it would be hard to provide a general purpose read API.

>     2. Why watermark is needed
>     3. Why watermark is designed in this way?

Agreed to both if this is an implementation

>     4. Why I have an extra sink function and operator class:
>        To avoid the effects/changes in the current interface. SinkFunctions and Operators are broadly used and I do not want to cause confusion or potential risks to our users.

Usually more code - especially duplicated code - leads to higher maintenance costs in the future. In particular, the operator is `@Internal`, so users should not use it directly and we do not give any guarantees on stableness. On the other hand, you modified the `@Public` `SinkFunction`, which needs to be carefully done (=have someone from SDK team look at it). I could see it being useful as a general extension independent of `KafkaShuffle`. Or we could actually add this method to a special `SinkFunctionWithWatermarks` or so.

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/SinkFunction.java:
```diff
@@ -52,6 +54,20 @@ default void invoke(IN value, Context context) throws Exception {
+	 * @throws Exception This method may throw exceptions. Throwing an exception will cause the operation
+	 *                   to fail and may trigger recovery.
+	 */
+	default void invoke(Watermark watermark) throws Exception {
```
If we use a custom operator for sink/source, we can avoid modifying the public API of `SinkFunction` at all. Stephan also talked about operators in the jira ticket.

curcur replied:
Mark as resolved since the entire API is re-organized.

SinkFunction is untouched in the new version.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaShuffleProducer.java:
```diff
@@ -0,0 +1,194 @@
+
+			try {
+				if (timestamp == null) {
+					dos.writeInt(TAG_REC_WITHOUT_TIMESTAMP);
```
My gut feeling is that we need to forward also some events, such as end of partition. 
I'm also not entirely sure how to deal with checkpoint barriers (see general remark).
In any case, we should use `StreamElementSerializer` instead of implementing it again.

curcur replied:
Mark as resolved since the entire API is re-organized.

StreamElementSerializer is discussed offline and summarized in the following comments.

curcur commented:
> So I had an offline discussion with Stephan to clarify the scope. Indeed, `KafkaShuffle` has been requested by users and serves as a bridge until we get fully persistent channels.
> 
> We both agree that it would have been nice to also support reading from non-Flink shuffles (a.k.a from any partitioned Kafka topic) by making the serializer pluggable at the composition level. Please have a look at `StreamElementSerializer` and see if we can use it. If that doesn't work for some reason, then I can live with a pure `KafkaShuffle` in the first iteration.
> 
> Implementation-wise, we are both a bit skeptical that an API change (to `SinkFunction`) is the best course as that requires more coordination and should have probably been triggered already if you want this feature in 1.11. Using custom operators would give you all freedom without that the need of coordiation. It would also avoid the changes to `KafkaProducer`/`KafkaConsumer` on the cost of replicating some logic.
> 
> Lastly, I have strong headaches on how checkpoints and savepoints are working with `KafkaShuffle`. I think for storing checkpoints and recovery in terms of fault tolerance, the approach is good as-is. However, for savepoints, we should probably ensure that no unconsumed data is still in lingering in the shuffle topic as that would translate to in-flight data. Hence, restoring from an old savepoint would completely screw up the data. At this point, we also need to ensure that the topic is purged (probably with some assertion). Not supporting going back in checkpoints should be save from current guarantees. Alternatively, we also need to implement some recovery logic for older check/savepoints that ignores "future" data somehow (so some higher level Kafka offset management).

Hey Arvid, thanks so much for the quick response!
I think you have several concerns about

1. why `StreamElementSerializer` can not be reused
2. why I have to have a different `KafkaProducer/KafkaConsumer`
3. Have a better way instead of changing `SinkFunction` (that's exactly my concern, and why I want to get early feedback, I am hesitating as well)
4. `save points`, which I do not completely get it.

For the first two, I have reasons; for the third one, I have concerns as well. For the forth, not completely sure I understand it correctly.
Do, do you have time to chat a bit on these four points on Monday?

BTW, I am not insisting to get this in 1.11. Instead, I really want to do it in the right way.

curcur commented:
# 1. Change to `SinkFunction`. 
By "coordination", do you mean that other changes are also made upon `SinkFunction` and may need coordination?

I've expected reviewers to have strong reactions to the API change, that's fine. But I am a bit confused about what is agreed/disagreed and what is a suggested better way, so let me try to clarify some of my thoughts and reason about why the API is changed in this way.

As suggested by Stephan, In the PR, I do have a custom operator `StreamShuffleSink` and a custom transformation in `SinkTransformation` for the custom operator. As Arvid mentioned in the previous reviews, there are a lot of code duplications between `StreamShuffleSink` and `StreamSink`. 
- That's true because they are very similar LOL, but we do want to provide a different operator to minimize the impact of changes on existing operators.
- Personally, I do not prefer to have multi-levels of extends/subclasses, especially if the superclass is not abstract. Multi-level extensions make code very difficult to read. You can not easily track what functions/members a class contains in a straightforward way, especially without a good IDE. 
- Come back to the duplication. There are in total 100 lines of code, with very simple logic. So personally I would prefer to trade these `100` lines of code for `readability`.

`SinkFunction` as its name, is the function invoked in the sink operator to provide a invoke function to handle record. `FlinkKafkaProducer` itself is a TwoPhaseCommitSinkFunction which implements `SinkFunction`.
If we really want to avoid changing `SinkFunction`, I can have a new interface and have the current TwoPhaseCommitSinkFunction implements the new interface. It should be safer than the current way, and also avoids conflicts if that's the concern.
 
Please let me know what do you think of this proposal.

# 2. `StreamElementSerializer`; 
I can not simply use `StreamElementSerializer` because the way watermark is stored/handled is different. In short, if multiple sink subtasks write to the same partition (sink), we need a way to decide the watermark in the source (downstream operator from the shuffle perspective). 
In the current netty shuffle service, we keep N channels and watermarks in each channel; while in this case, data and watermarks have been merged when writing to partitions.

Please refer to Jira FLINK-15670 for discussion about watermark: https://issues.apache.org/jira/browse/FLINK-15670

You can start from [here](https://issues.apache.org/jira/browse/FLINK-15670?focusedCommentId=17053232&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17053232)

It includes my original thoughts, proposals and Stephan's enhanced version.

# 3. "Checkpoints" and "Savepoints"
As far as I know, savepoints are very similar to checkpoints except that savepoints are more or less user-faced. That says user can trigger a replay based on save points. I guess I can kind of understanding why you are saying "restoring from an old savepoint would completely screw up the data". It is true if you think of this problem from a global snapshotting and global failover perspective. 

However, let's step back and think of why we we want to have the persistent shuffle in the first place. If data is persisted, you do not really need to replay the calculation again. Persistency is to unleash the constraints between upstream and downstream.

For your concern, we do not need to do a global replay as well. We can simply do a regional replay. If there is any constraints in implementation, we can disable it for now. In the long term, I do not see it is a problem.

But, Maybe I misunderstand you :-)

Sorry to write this much again. These are all good questions that I think is more clear to answer in a systematic way :-). Let's chat in details this afternoon.

curcur commented:
Updates to commit "Kafka Shuffle and Test Case"

1. add log4j2 file for better testing
2. I found the original tests are occasionally failing when testing for a large number of data. The reason is the default property "auto.offset.reset" is set to "latest". That means if the consumer starts later than the producer starting to writing data, then I am going to lose some data occasionally.
So, I switched to use KafkaTestEnvironment.getStandardProperties which set the auto.offset.reset = earliest

AHeise commented:
So another round concerning API changes with @pnowojski . He shares the concerns on extending `SinkFunction` and `TwoPhaseCommitFunction` for something that could be deemed an implementation detail.

So far we have discussed going either the `SinkFunction` or the custom operator route. Piotr actually proposed to do both at once:
* Instead of some generic `StreamShuffleSink`, have a very specific `KafkaShuffleSink` that extends `StreamSink` and pretty much just overrides the watermark processing.
* Also implement the `FlinkKafkaShuffleProducer` in the way that you have. But instead of overriding the newly added watermark functions in `SinkFunction` and `TwoPhaseCommitFunction`, just provide these methods explicitly.
* The `KafkaShuffleSink` will then call this function directly in `processWatermark`.

This approach has several advantages:
* No changes to Public API (`SinkFunction` and `TwoPhaseCommitFunction`).
* Full control over watermark handling without forking out some general concept.
* No code replication in operator and `SinkFunction`.
* Little changes to your current approach (most changes concern glue code).

The only downside is that functionality is a bit split on two entities, but that's also true for the current approach. With this approach, you could even have both entities in the same class (e.g., make `FlinkKafkaShuffleProducer` an inner class of `KafkaShuffleSink`).

Another advantage is that all implementation work can be done in `flink-connector-kafka`, thus solving the cyclic dependency issue that you raised.
Further, you can decrease the visibility to all `KafkaProducer` and `KafkaConsumer` changes from `protected` to `package-private`, making it more obvious that we don't see these things as public API.

curcur commented:
Hey, @AHeise  and @pnowojski. Thanks so much for your time and efforts, really appreciate!

I think this is a fair and practical workaround. My first intention was that "operator" is put with the API module, and function implementation is put with the Kafka connector module. But now since the operator is tailored specifically for new KafkaShuffleFunction, then it is no need to put the custom operator in the API module anymore. 

If I wrap everything in the `flink-connector-kafka`, I agree there should not be any cyclic dependency issue. 

I will do the updates, and thank you both again!

curcur commented on flink-core/src/main/java/org/apache/flink/util/PropertiesUtil.java:
```diff
@@ -108,6 +109,27 @@ public static boolean getBoolean(Properties config, String key, boolean defaultV
+	 * @param config Properties to be flatten
+	 * @return Properties without defaults; all properties are put in the first-level
+	 */
+	public static Properties flatten(Properties config) {
```
> On a second thought, wouldn't it make more sense to provide a correct `propsToMap` implementation? If it's only used in KafkaProducer, then we could fix it there. If not, I'd consider that function more useful than this `flatten`.

this name is still confusing since it is a Properties, not a map.

The flattened properties are actually used in the Kafka client lib, not that easy to fix.

curcur replied:
> Should be covered with one test case.

sure, will add one later.

curcur commented on flink-core/src/main/java/org/apache/flink/util/PropertiesUtil.java:
```diff
@@ -19,6 +19,7 @@
 
```
Oh, yep, used as a HashTable, translated to a HashMap. Thanks :-)

curcur commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamShuffleSink.java:
```diff
@@ -0,0 +1,104 @@
+ * @param <IN>
+ */
+@Internal
+public class StreamShuffleSink<IN> extends AbstractUdfStreamOperator<Object, SinkFunction<IN>>
```
Mark this as resolved since the entire API part has been re-organized.

curcur commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java:
```diff
@@ -1333,6 +1334,32 @@ public ExecutionConfig getExecutionConfig() {
+	 * 			The object containing the sink's invoke function for both the element and watermark.
+	 * @return	The closed DataStream.
+	 */
+	public DataStreamSink<T> addSinkShuffle(SinkFunction<T> sinkFunction) {
```
Mark as resolved since the entire API is re-organized.

This method is wrapped inside `FlinkKafkaShuffle`

curcur commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkTransformation.java:
```diff
@@ -67,6 +68,22 @@ public SinkTransformation(
+	 * @param operator The sink shuffle operator
+	 * @param parallelism The parallelism of this {@code SinkTransformation}
+	 */
+	public SinkTransformation(
```
Mark as resolved since the entire API is re-organized.

curcur commented:
@AHeise and @pnowojski , I have updated the PR with respect to the API change suggested above. The API change is listed as the fourth commit "[FLINK-15670][connector] Kafka Shuffle API Part". The updated PR also includes amends suggested by Arvid in previous reviews. I left a reply for comments not yet resolved.  

Now, most of the code change is wrapped in the package "org.apache.flink.streaming.connectors.kafka.shuffle" to avoid effects on other parts.

1. I think now the only thing a bit struggling is "TwoPhaseCommitFunction". If I do not provide a watermark entry in that abstract function, I will end up exposing "currentTransactionHolder" as well as "TransactionHolder#handle". Your call.

2. I divide the "persistentKeyBy" to two functions: "writeKeyBy" and "readKeyBy". Hence people can call "readKeyBy" directly to reuse the written data. You can take a look at the API to see whether the changes make sense.

Thanks in advance!!

AHeise commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java:
```diff
@@ -264,7 +264,7 @@ public FlinkKafkaConsumerBase(
 	 * @param offsetCommitMode offset commit mode
 	 */
-	static void adjustAutoCommitConfig(Properties properties, OffsetCommitMode offsetCommitMode) {
+	public static void adjustAutoCommitConfig(Properties properties, OffsetCommitMode offsetCommitMode) {
```
protected?

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,344 @@
+
+			subtaskWatermark.put(newWatermark.subtask, newWatermark.watermark);
+
+			if (subtaskWatermark.values().size() < producerParallelism) {
+				return Optional.empty();
+			}
```
What happens if one partition has ended and we receive no watermarks anymore? Are the watermarks of the other partitions still propagated properly? Almost feels like using `StatusWatermarkValve` would be handy.

curcur replied:
Yes, they are very similar at first glance. But my feeling is `StatusWatermarkValve` is an overkill since it also includes StreamStatus. And the pick up of watermark depends on the stream status.

- If one partition has ended and we receive no watermarks anymore, are the watermarks of the other partitions still propagated properly?

As long as all subtasks contain watermarks, the watermark progress properly. The only problem is if one of the subtask does not produce watermark. But that's a common question to find a min/max watermark amongst all channels. 

Data is propogates fine.

AHeise replied:
I'd assume you need to propagate EndOfPartitionEvents downstream and adjust `producerParallelism` accordingly.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,229 @@
+/**
+ * Use Kafka as a persistent shuffle by wrapping a Kafka Source/Sink pair together.
+ */
+class FlinkKafkaShuffle {
```
If this is API, I guess it should be `public` and either `@PublicEvolving` or `@Experimental`.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,229 @@
+	 * @param <T> input type
+	 */
+	static <T> KeyedStream<T, Tuple> persistentKeyBy(
+		DataStream<T> inputStream,
```
nit: indent.

curcur replied:
I will do a final pass later.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,229 @@
+	 * @param fields key positions from inputStream
+	 * @param <T> input type
+	 */
+	static <T> KeyedStream<T, Tuple> persistentKeyBy(
```
`public`

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,229 @@
+	 *
+	 * @param inputStream input stream to the kafka
+	 * @param topic kafka topic
+	 * @param producerParallelism parallelism of producer
+	 * @param numberOfPartitions number of partitions
```
Shouldn't that be the same?

curcur replied:
producerParallelism can be different from numberOfPartitions

AHeise replied:
Yes, but what is the benefit? I guess it can be used to fan-in/fan-out at the source level?

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java:
```diff
@@ -0,0 +1,309 @@
+	 * Producer Parallelism = 1; Kafka Partition # = 1; Consumer Parallelism = 1.
+	 * To test no data is lost or duplicated end-2-end with the default time characteristic: ProcessingTime
+	 */
+	@Test(timeout = 30000L)
```
Instead of setting timeout to all methods, I'd go with a JUnit rule:
```
	@Rule
	public final Timeout timeout = Timeout.builder()
			.withTimeout(30, TimeUnit.SECONDS)
			.build();
```

and then only use `@Test` on the tests. That's easier to maintain when we need to increase the timeout on azure.

curcur replied:
that's a good one :-)

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java:
```diff
@@ -0,0 +1,309 @@
+	 * Producer Parallelism = 1; Kafka Partition # = 1; Consumer Parallelism = 1
+	 * To test no data is lost or duplicated end-2-end
+	 */
+	private void simpleEndToEndTest(String topic, int elementCount, TimeCharacteristic timeCharacteristic)
```
We use end2end in a different context, where we use a complete Flink distribution to execute the test. 

I'd simply call it `testKafkaShuffle` to avoid any misunderstanding.

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java:
```diff
@@ -0,0 +1,309 @@
+		final int numberOfPartitions = 3;
+		final int producerParallelism = 2;
+
+		createTestTopic(topic, numberOfPartitions, 1);
+
+		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		env.setParallelism(producerParallelism);
+		env.setRestartStrategy(RestartStrategies.noRestart());
+		env.setStreamTimeCharacteristic(EventTime);
+
+		DataStream<Tuple3<Integer, Long, String>> source =
+			env.addSource(new KafkaSourceFunction(elementCount)).setParallelism(producerParallelism);
+
+		DataStream<Tuple3<Integer, Long, String>> input = (timeCharacteristic == EventTime) ?
+			source.assignTimestampsAndWatermarks(new PunctuatedExtractor()).setParallelism(producerParallelism) : source;
+
+		// ------- Write data to Kafka partition basesd on FlinkKafkaPartitioner ------
+		Properties properties = kafkaServer.getStandardProperties();
+
+		KeyedStream<Tuple3<Integer, Long, String>, Tuple> keyedStream = FlinkKafkaShuffle
+			.persistentKeyBy(input, topic, producerParallelism, numberOfPartitions, properties, 0);
+		keyedStream
+			.process(new PartitionValidator(keyedStream.getKeySelector(), numberOfPartitions, topic))
+			.setParallelism(numberOfPartitions)
+			.map(new ElementCountNoMoreThanValidator(elementCount * producerParallelism)).setParallelism(1)
+			.map(new ElementCountNoLessThanValidator(elementCount * producerParallelism)).setParallelism(1);
+
+		tryExecute(env, "KafkaShuffle partition assignment test");
+
+		deleteTestTopic(topic);
```
Extract to avoid duplicate code with `simpleEndToEndTest`.

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java:
```diff
@@ -0,0 +1,309 @@
+	}
+
+	private static class ElementCountNoMoreThanValidator
+		implements MapFunction<Tuple3<Integer, Long, String>, Tuple3<Integer, Long, String>> {
```
nit: also double-indent.

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java:
```diff
@@ -0,0 +1,309 @@
+
+		@Override
+		public void processElement(
+			Tuple3<Integer, Long, String> in, Context ctx, Collector<Tuple3<Integer, Long, String>> out)
```
nit: chop args

curcur replied:
I will do a pass on it.

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java:
```diff
@@ -0,0 +1,309 @@
+/**
+ * Simple End to End Test for Kafka.
+ */
+public class KafkaShuffleITCase extends KafkaConsumerTestBase {
```
The implemented tests are really good. I miss two cases though:
* Out of order events (add randomness to source timestamp)
* Any failure and recovery tests. See https://github.com/apache/flink/blob/f239d680e9b8f3f5ace621b7806e0bb7e14d3fdd/flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointITCase.java for a possible approach.

curcur replied:
Both failover cases and watermark cases are added.

But I feel I still need some more tests

1. one on the coumser side of watermarks
2. one on the producer side failover.

curcur commented:
Thanks so much @AHeise !

The watermark (from Producer side) is almost done, need to do some clean-up though.

Will update the PR and also address the comments once the remaining IT tests are done.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffleProducer.java:
```diff
@@ -0,0 +1,197 @@
+	private final int numberOfPartitions;
+
+	FlinkKafkaShuffleProducer(
+		String defaultTopicId,
```
I will do a pass on it....

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffleProducer.java:
```diff
@@ -0,0 +1,197 @@
+	/**
+	 * Flink Kafka Shuffle Serializer.
+	 */
+	public static final class KafkaSerializer<IN> implements Serializable {
+		public static final int TAG_REC_WITH_TIMESTAMP = 0;
+		public static final int TAG_REC_WITHOUT_TIMESTAMP = 1;
+		public static final int TAG_WATERMARK = 2;
```
This is a good idea.

Actually I find the SerDe part might be able to be reused by other persistent storage, not just Kafka, so it is necessary to exact Ser/De to the top level.

Let's see whether I have time to extract these parts out before

curcur replied:
As discussed offline, maybe extracting it next time when a similar persistent channel is needed.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,344 @@
+					for (ConsumerRecord<byte[], byte[]> record : partitionRecords) {
+						final KafkaShuffleElement<T> element = deserializer.deserialize(serializer, record);
+
+						// TODO: do we need to check the end of stream if reaching the end watermark?
```
OK, let's discuss these offline.

On the other hand, my question is actually whether I need to pass the MAX watermark downstream?

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,344 @@
+		}
+
+		boolean isWatermark() {
+			return getClass() == KafkaShuffleWatermark.class;
+		}
+
+		KafkaShuffleRecord<T> asRecord() {
+			return (KafkaShuffleRecord<T>) this;
+		}
+
+		KafkaShuffleWatermark asWatermark() {
+			return (KafkaShuffleWatermark) this;
+		}
+	}
+
+	private static class KafkaShuffleWatermark<T> extends KafkaShuffleElement<T> {
+		final int subtask;
+		final long watermark;
+
+		KafkaShuffleWatermark(int subtask, long watermark) {
+			this.subtask = subtask;
+			this.watermark = watermark;
+		}
+	}
+
+	private static class KafkaShuffleRecord<T> extends KafkaShuffleElement<T> {
+		final T value;
+		final Long timestamp;
+
+		KafkaShuffleRecord(T value) {
+			this.value = value;
+			this.timestamp = null;
+		}
+
+		KafkaShuffleRecord(long timestamp, T value) {
+			this.value = value;
+			this.timestamp = timestamp;
+		}
+	}
+
```
As stated above, the SerDe part can be generalized. it is not just Kafka.

curcur replied:
do it if having time.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffleProducer.java:
```diff
@@ -0,0 +1,197 @@
+			throw new RuntimeException("Fail to assign a partition number to record");
+		}
+
+		ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(
```
I think during my first test, the timestamp read from ConsumerRecord is always the timestamp when it was created. 
But i did not spend time looking into this because as stated above, the SerDe part can be generalized. it is not just Kafka. So i would like to make the SerDe less dependant on Kafka.

AHeise replied:
Okay seems plausible.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,344 @@
+			if (tag == TAG_REC_WITHOUT_TIMESTAMP) {
+				return new KafkaShuffleRecord<>(serializer.deserialize(dis));
+			} else if (tag == TAG_REC_WITH_TIMESTAMP) {
+				return new KafkaShuffleRecord<>(LongSerializer.INSTANCE.deserialize(dis), serializer.deserialize(dis));
```
same answer above.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,344 @@
+							}
+						} else if (element.isWatermark()) {
+							final KafkaShuffleWatermark watermark = element.asWatermark();
+							Optional<Watermark> newWatermark = watermarkHandler.checkAndGetNewWatermark(watermark);
+							newWatermark.ifPresent(sourceContext::emitWatermark);
```
This is a rewrite of {@link AbstractFetcher#emitRecordWithTimestamp}

         if (record != null) {
			if (timestampWatermarkMode == NO_TIMESTAMPS_WATERMARKS) {
				// fast path logic, in case there are no watermarks generated in the fetcher

				// emit the record, using the checkpoint lock to guarantee
				// atomicity of record emission and offset state update
				synchronized (checkpointLock) {
					sourceContext.collectWithTimestamp(record, timestamp);
					partitionState.setOffset(offset);
				}
			} else if (timestampWatermarkMode == PERIODIC_WATERMARKS) {
				emitRecordWithTimestampAndPeriodicWatermark(record, partitionState, offset, timestamp);
			} else {
				emitRecordWithTimestampAndPunctuatedWatermark(record, partitionState, offset, timestamp);
			}
		} else {
			// if the record is null, simply just update the offset state for partition
			synchronized (checkpointLock) {
				partitionState.setOffset(offset);
			}
		}

curcur replied:
If you go through `emitRecordWithTimestampAndPunctuatedWatermark` 
for exmaple

records are emitted under checkpointLock
watermarks are emitted through updateMinPunctuatedWatermark(newWatermark), not within the lock.

This is understandable in the current case, since watermark is only a indicator, not strictly part of a `checkpoint`. That's why.

AHeise replied:
Yes, you are not modifying state, so all good.

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java:
```diff
@@ -0,0 +1,722 @@
+	 */
+	@Test
+	public void testSimpleProcessingTime() throws Exception {
+		testKafkaShuffle("test_simple_" + ProcessingTime.toString(), 200000, ProcessingTime);
```
Seems like you add the `timeCharacteristic` always to the topic name. If so, then you can move it into the respective test implementation function to make the invocation cleaner.

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java:
```diff
@@ -0,0 +1,722 @@
+						fail("Watermarks should not be generated in the case of ProcessingTime");
+						break;
+					case IngestionTime:
+						break;
+					case EventTime:
+						KafkaShuffleWatermark watermark = element.asWatermark();
```
Test if watermarks are really monotonic.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,367 @@
+	 */
+	private static class WatermarkHandler {
+		private final int producerParallelism;
+		private final Map<Integer, Long> subtaskWatermark;
```
long[]?

curcur commented:
Fix the problem Failover tests:
1. For two-phase commit, if one of the subtasks is FINISHEDthen the whole graph aborts because checkpoint can not complete. The solution is using unbounded sources instead.

2. Add states (snapshot/restore) for validation checks when enabling checkpoints.

curcur commented:
Stable ITTests, finally...

curcur commented:
@AHeise 
I think I've addressed most of the comments, and all the tests are stable now.

I also did one pass to address the java docs and indent problem.

Missing java docs generates a check-style error, so it is relatively easy to check.  For the indent problem, I've tried my best. Hope I did not miss places.

I think I am still missing one "must-do", randomize watermarks.

I will think of one tonight and add one tomorrow. It should be fairly easy (well compared to the failover IT tests).

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,231 @@
+		transformation.setParallelism(producerParallelism);
+	}
+
+	// A better place to put this function is DataStream; but put it here for now to avoid changing DataStream
```
I wouldn't mind as this only changes internals.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/StreamKafkaShuffleSink.java:
```diff
@@ -0,0 +1,44 @@
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		super.processWatermark(mark);
+		this.currentWatermark = mark.getTimestamp();
```
this is already done by `super`. Afaik you then also don't need the change in StreamSink.

curcur replied:
That's a good catch! Thanks!!

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStreamSink.java:
```diff
@@ -35,9 +35,9 @@
 
-	@SuppressWarnings("unchecked")
 	protected DataStreamSink(DataStream<T> inputStream, StreamSink<T> operator) {
-		this.transformation = new SinkTransformation<T>(inputStream.getTransformation(), "Unnamed", operator, inputStream.getExecutionEnvironment().getParallelism());
```
Unrelated changes, please revert or pull into a separate hotfix if you deem it necessary (I don't).

curcur replied:
I will revert it.

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java:
```diff
@@ -0,0 +1,412 @@
+	 * To test no data is lost or duplicated end-2-end
+	 */
+	private void testKafkaShuffle(
+			String prefix, int numElementsPerProducer, TimeCharacteristic timeCharacteristic) throws Exception {
```
chop

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java:
```diff
@@ -0,0 +1,412 @@
+
+		final StreamExecutionEnvironment env = createEnvironment(producerParallelism, timeCharacteristic);
+		createKafkaShuffle(
+			env, topic, numElementsPerProducer, producerParallelism, timeCharacteristic, numberOfPartitions)
```
chop

curcur commented on flink-core/src/main/java/org/apache/flink/util/PropertiesUtil.java:
```diff
@@ -108,6 +109,27 @@ public static boolean getBoolean(Properties config, String key, boolean defaultV
 
+	/**
+	 * Flatten a recursive {@link Properties} to a first level property map.
+	 * In some cases, {KafkaProducer#propsToMap} for example, Properties is used purely as a HashTable
```
Yep, that's why I removed the link :-)

Thanks!

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:
```diff
@@ -210,7 +210,7 @@
 	/**
```
I like "by heart", hahaha

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:
```diff
@@ -210,7 +210,7 @@
 	 * The name of the default topic this producer is writing data to.
 	 */
-	private final String defaultTopicId;
+	protected final String defaultTopicId;
```
It was fine to use "package-private" previously, but I have moved all the KafkaShuffle related classes to a separate package. So, it is not sufficient any more.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffleProducer.java:
```diff
@@ -0,0 +1,195 @@
+		}
+
+		ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(
+			defaultTopicId, partitionIndex, timestamp, null, kafkaSerializer.serializeRecord(next, timestamp));
```
OK, I will do another pass.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffleProducer.java:
```diff
@@ -0,0 +1,195 @@
+
+			try {
+				if (timestamp == null) {
+					dos.writeInt(TAG_REC_WITHOUT_TIMESTAMP);
```
It should be, but again, this is similar to the tag handling of 
`StreamElementSerializer`

Would you mind keeping as it is for now?

curcur replied:
I will do this.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,380 @@
+ * Fetch data from Kafka for Kafka Shuffle.
+ */
+@Internal
+public class KafkaShuffleFetcher<T> extends AbstractFetcher<T, TopicPartition> {
```
That's a good question. 

It is because what I really override is fetcher's method (how fetcher fetch records, deserialize records and emit watermarks). That's why I have to subclass Fetcher anyway.

I tried different ways, there is one way to avoid duplicated code, but needs to add a default constructor in one of consumer or fetcher base classes.

I think it is unsafe to do it, and that's why it is ended up like this.

curcur replied:
me answer why.

curcur replied:
removed the duplicated code.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffleConsumer.java:
```diff
@@ -0,0 +1,90 @@
+	private final TypeSerializer<T> serializer;
+	private final int producerParallelism;
+
+	FlinkKafkaShuffleConsumer(String topic, TypeInformationSerializationSchema<T> schema, Properties props) {
```
That's not easy, because I have to call the super constructor of FlinkKafkaConsumer.

Schema is needed for the constructor of FlinkKafkaConsumer

super(topic, schema, props);

curcur replied:
I guess the main point here is to avoid exposing `TypeInformationSerializationSchema` to users. `FlinkKafkaShuffleConsumer` is an internal function, and is wrapped under KafkaShuffle. So I guess it is fine to keep TypeInformationSerializationSchema<T> it here.
The schema is needed to call the right FlinkKafkaConsumer constructor.
The shema is never used, can be `null`, but `null` confuses the compiler.

curcur replied:
I will mark this resolved for now.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,231 @@
+	 * Write to and read from a kafka shuffle with the partition decided by keys.
+	 * Consumers should read partitions equal to the key group indices they are assigned.
+	 * The number of partitions is the maximum parallelism of the receiving operator.
+	 * This version only supports numberOfPartitions = consumerParallelism.
```
Do you mean specifically saying "producerParallelism != numberOfPartitions?"

If think this is keyBy through a persistent channel, the producer parallelism does not matter with the max key group size? I kind of feeling confusing if adding such a comment.

For config changes, yes, probably in the next step; Overall, this is only a very first version, allowing users to set parallelism when writing pipelines might be enough.

AHeise replied:
> the producer parallelism does not matter with the max key group size?

If it doesn't matter, why can it be set by the user? Please give users some guidance on how to use your function. How should they know how it's implemented internally? You could also add it as a class comment to outline the approach.

In general, there are quite a bit of comments and variable names referring to producer, but there is no explanation what the producer is in KafkaShuffle. I'd probably harmonize source/consumer and sink/producer and only use either word.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,380 @@
+		public KafkaShuffleElement<T> deserialize(TypeSerializer<T> serializer, ConsumerRecord<byte[], byte[]> record)
+			throws Exception {
+			byte[] value = record.value();
+			dis.setBuffer(value);
```
That's actually my question, why?

I have the same problem for dos, but it seems working here?

curcur commented:
address comments and did another pass on java docs and formats.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,248 @@
+	}
+
+	/**
+	 * Writes to a kafka shuffle with the partition decided by keys.
```
What does "Writes to a kafka shuffle" mean? I think you need to add more explanation.
This function also is only meaningful with `readKeyBy` so add a `@see`.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,248 @@
+	/**
+	 * Reads data from a Kafka Shuffle.
+	 *
+	 * <p>Consumers should read partitions equal to the key group indices they are assigned.
+	 * The number of partitions is equivalent to the key group sizes. Each consumer task reads from
+	 * one or multiple partitions. Any two consumer tasks can not read from the same partition.
+	 * Hence, the maximum parallelism of the receiving operator is the number of partitions.
+	 * This version only supports numberOfPartitions = consumerParallelism
```
This is a good comment. Something like that should exist also for the other public API methods.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,248 @@
+	}
+
+	/**
+	 * Reads data from a Kafka Shuffle.
```
Reads data from a Kafka Shuffle previously written by `writeKeyBy`.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,248 @@
+	public static <T, K> KeyedStream<T, K> readKeyBy(
+			String topic,
+			StreamExecutionEnvironment env,
+			TypeInformationSerializationSchema<T> schema,
```
I'd go with `TypeInformation` here.  `TypeInformationSerializationSchema` is rather technical and can be easily derived from `TypeInformation`.

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleExactlyOnceITCase.java:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	private StreamExecutionEnvironment createEnvironment(
+			int producerParallelism, TimeCharacteristic timeCharacteristic) {
```
chop

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,379 @@
+	 * - watermark timestamp
+	 */
+	@VisibleForTesting
+	public static class KafkaShuffleWatermark<T> extends KafkaShuffleElement<T> {
```
`KafkaShuffleWatermark` is always used as a raw type, so please get rid of the type parameter. 

I also think that `KafkaShuffleElement<T>` should be without type parameter and only `asRecord` has it.
```
	public <T> KafkaShuffleRecord<T> asRecord() {
			return (KafkaShuffleRecord<T>) this;
	}
```

curcur replied:
Yes, good catch.

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleExactlyOnceITCase.java:
```diff
@@ -0,0 +1,204 @@
+		@Override
+		public Integer map(Tuple3<Integer, Long, Integer> element) throws Exception {
+			int addedInteger = element.f0 * producerParallelism + element.f2;
+			System.out.println("<" + element.f0 + "," + element.f2 + "> " + addedInteger);
```
Remove or translate into LOG.debug

AHeise commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleTestBase.java:
```diff
@@ -0,0 +1,213 @@
+ * Base Test Class for KafkaShuffle.
+ */
+public class KafkaShuffleTestBase extends KafkaConsumerTestBase {
+	static final long INIT_TIMESTAMP = System.currentTimeMillis();
```
Can't we just use a random static number here? I was thinking that tests results might be easier to compare if it's not a real time stamp.

curcur replied:
why it is easier if not a real timestamp? I can change it back to a static number, but feeling this might be more realistic.

This change is originally motivated by debugging the Kafka Server. I thought some data is trimmed because of the timestamp is too old. But it is not related.

AHeise replied:
As I have said, if I want to compare two shuffle topics created by two commits to track down some bug, the changed timestamp will actually be a big PITA.
Tests don't need to be realistic in the sense of data, but should be realistic in terms of workload (or queries).

curcur commented:
Add watermark randomness test and incremental test from the consumer side.

I think by far both the value and watermark part are checked.

curcur commented:
> Test code already looks good (expecting the watermark tests to be implemented).
> 
> Production code needs polishing:
> 
> * API Javadoc should really help non-experts.
> * Serializers should probably more use the existing abstractions like `KafkaDeserializationSchema` and not change/exhibit `TypeInformationSerializationSchema`

This is difficult as what KafkaDeserializationSchema is providing

T deserialize(ConsumerRecord<byte[], byte[]> record) throws Exception;

and what I need, a TypeSerialer that can deserialize type by type.

If I am asking for a KafkaDeserializationSchema, that's basically asking the user to write the deserializer themselves.

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,380 @@
+						records.records(partition.getKafkaPartitionHandle());
+
+					for (ConsumerRecord<byte[], byte[]> record : partitionRecords) {
+						final KafkaShuffleElement<T> element = deserializer.deserialize(serializer, record);
+
+						// TODO: do we need to check the end of stream if reaching the end watermark?
+
+						if (element.isRecord()) {
+							// timestamp is inherent from upstream
+							// If using ProcessTime, timestamp is going to be ignored (upstream does not include timestamp as well)
+							// If using IngestionTime, timestamp is going to be overwritten
+							// If using EventTime, timestamp is going to be used
+							synchronized (checkpointLock) {
+								KafkaShuffleRecord<T> elementAsRecord = element.asRecord();
+								sourceContext.collectWithTimestamp(
+									elementAsRecord.value,
+									elementAsRecord.timestamp == null ? record.timestamp() : elementAsRecord.timestamp);
+								partition.setOffset(record.offset());
+							}
+						} else if (element.isWatermark()) {
+							final KafkaShuffleWatermark watermark = element.asWatermark();
+							Optional<Watermark> newWatermark = watermarkHandler.checkAndGetNewWatermark(watermark);
+							newWatermark.ifPresent(sourceContext::emitWatermark);
+						}
```
Me Answer why.

curcur commented:
As discussed offline, three things remaining to do:

1. Ser/Der Format Change
2. avoid duplicated code
3. Use TypeSerializer instead of TypeInformationSerializationSchema

From the priority wise, I think 1 > 2 > 3.

curcur commented:
Resolved 1. Ser/Der Format Change

curcur commented:
Resolved 2. Avoid duplicated code

1. `KafkaShuffleFetcher` now extends `KafkaFetcher` instead of `AbstractFetcher`
2. I also did a little clean up on the code of `KafkaFetcher`. I know it is not quite relevant, but I guess you will think it is necessary as well.
3. The current "ending of stream" logic in KafkaFetcher a bit strange: if any partition has a record signaled as "END_OF_STREAM", the fetcher will stop running. Notice that the signal is coming from the deserializer, which means from Kafka data itself. But it is possible that other topics and partitions still have data to read. Finishing reading Partition0 can not guarantee that Partition1 also finishes. It is possible I misunderstand what an "END_OF_STREAM" signal means. 
4. All Kafka related tests passed after the change.

curcur commented:
Resolved 3. the last one.

remove getTypeSerializer() from TypeInformationSerializationSchema

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaFetcher.java:
```diff
@@ -192,6 +176,29 @@ protected String getFetcherName() {
+			KafkaTopicPartitionState<TopicPartition> partition) throws Exception {
+
+		for (ConsumerRecord<byte[], byte[]> record : partitionRecords) {
+			deserializer.deserialize(record, kafkaCollector);
+
+			// emit the actual records. this also updates offset state atomically and emits
+			// watermarks
+			emitRecordsWithTimestamps(
+				kafkaCollector.getRecords(),
+				partition,
+				record.offset(),
+				record.timestamp());
+
+			if (kafkaCollector.isEndOfStreamSignalled()) {
+				// end of stream signaled
+				running = false;
+				break;
+			}
```
Much better now, but I would have just extracted this inside of the for loop, since the loop itself is also the same.

curcur replied:
> Much better now, but I would have just extracted this inside of the for loop, since the loop itself is also the same.

It would be a bit tricky because it has a for-loop break; I have to change the logic inside. At this time, I would prefer to make more safe changes...

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,291 @@
+			autoWatermarkInterval,
+			userCodeClassLoader,
+			taskNameWithSubtasks,
+			null,
```
passing null on a parameter not annotated with `Nullable` is quite hacky and can fail once someone adds a non-null check (which should be there). Maybe use the same trick of anonymous function as you did in `KafkaProducer`? If not, really extract the `AbstractKafkaFetcher`.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java:
```diff
@@ -0,0 +1,379 @@
+		}
+
+		@VisibleForTesting
+		public KafkaShuffleElement<T> deserialize(TypeSerializer<T> serializer, ConsumerRecord<byte[], byte[]> record)
```
Doesn't seem resolved.

curcur commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java:
```diff
@@ -264,7 +264,7 @@ public FlinkKafkaConsumerBase(
 	 * @param properties - Kafka configuration properties to be adjusted
 	 * @param offsetCommitMode offset commit mode
```
Do you mean remove "-" ?
I did not change this code?

curcur commented:
Hey @AHeise , thank you so much again!

I believe I've addressed most of the minor issues mentioned in the previous reviews. Please remind me if I miss anything but for the one "extracted this inside of the for loop". I've left an offline msg to you to explain why I think it is better not to do it right now.

FLINK-17307 changes the KafkaFetcher to emit the output in batch (once collector per partition), but mine is using once per record (emitting record and watermark in a different way).

You can see they have logical differences. It might worth changing to batch emit as well, but might be two collectors, one for watermarks and the other one for records.

I am not saying the change is difficult, but I am learning to be conservative right now and I've already listed it as a follow-up: https://issues.apache.org/jira/browse/FLINK-15670.

What do you think?

I am going to work on the Java docs for the public API now. If you have any comments on the code, please let me know.

curcur commented:
Javadocs for public API has been updated as well.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,351 @@
+	}
+
+	/**
+	 * Uses Kafka as a message bus to persist keyBy shuffle.
```
copy long explanation from above.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,351 @@
+	}
+
+	/**
+	 * The write side of {@link FlinkKafkaShuffle#persistentKeyBy}.
+	 *
+	 * <p>This function contains a {@link FlinkKafkaShuffleProducer} to shuffle and persist data in Kafka.
+	 * {@link FlinkKafkaShuffleProducer} uses the same key group assignment function
+	 * {@link KeyGroupRangeAssignment#assignKeyToParallelOperator} to decide which partition a key goes.
+	 * Hence, each producer task can potentially write to each Kafka partition based on the key.
+	 * Here, the number of partitions equals to the key group size.
+	 * In the case of using {@link TimeCharacteristic#EventTime}, each producer task broadcasts each watermark
+	 * to all of the Kafka partitions to make sure watermark information is propagated properly.
+	 *
+	 * <p>Attention: make sure kafkaProperties include
+	 * {@link FlinkKafkaShuffle#PRODUCER_PARALLELISM} and {@link FlinkKafkaShuffle#PARTITION_NUMBER} explicitly.
+	 * {@link FlinkKafkaShuffle#PRODUCER_PARALLELISM} is the parallelism of the producer.
+	 * {@link FlinkKafkaShuffle#PARTITION_NUMBER} is the number of partitions.
+	 * They are not necessarily the same and allowed to be set independently.
```
very good

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,351 @@
+	 * {@link FlinkKafkaShuffle#PARTITION_NUMBER} is the number of partitions.
+	 * They are not necessarily the same and allowed to be set independently.
+	 *
+	 * @see FlinkKafkaShuffle#persistentKeyBy
+	 * @see FlinkKafkaShuffle#readKeyBy
```
usually added at the very top of a javadoc

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,351 @@
+	}
+
+	/**
+	 * The write side of {@link FlinkKafkaShuffle#persistentKeyBy}.
```
copy javadoc

curcur commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java:
```diff
@@ -0,0 +1,351 @@
+ * <p>1). Partitioning is done through {@link FlinkKafkaShuffleProducer}. {@link FlinkKafkaShuffleProducer} decides
+ * 			which partition a key goes when writing to Kafka
+ *
+ * <p>2). Shuffle data can be reused through {@link FlinkKafkaShuffle#readKeyBy}, as shown in the example above.
```
I mean shuffle can be reused/re-read. I guess streaming data is not reserved?

curcur replied:
I can change reuse -> read again?

curcur commented:
Thank you both so much! This is my very first feature and it means a lot to me. I am especially impressed by the level of code integrity @AHeise insists. The code quality is significantly improved! Really appreciate, and thank you again!

curcur commented:
BTW, I have updated the javaDocs, and pushed.

pnowojski commented:
LGTM, private azure test passed % end to end tests timeout, but this couldn't be possibly caused by the latest update - previous version differed only in a java doc and was green. Merging.

rmetzger commented:
Note: changes merged through this PR broke master: https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=1686&view=results

rmetzger commented:
Change has been reverted https://github.com/apache/flink/commit/bf58725e7ad80f276a458007a2d5b890d8ffc4f5

curcur commented:
It is because of commit of unifying the watermark Strategy committed last night...

curcur commented:
The PR is reverted because of conflicts with
[FLINK-17659] Rework WatermarkStrategy, add Suppliers for TimestampAssigner/WatermarkGenerator

committed last night.

File the new PR: https://github.com/apache/flink/pull/12219

rmetzger commented:
Looks like this PR introduced a test instability? https://issues.apache.org/jira/browse/FLINK-17912
